dev/codehost_testing/scenario.go

```diff
@@ -0,0 +1,213 @@
+package codehost_testing
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/google/uuid"
+
+	"github.com/sourcegraph/sourcegraph/dev/codehost_testing/config"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+type action struct {
+	name     string
+	apply    func(context.Context) error
+	teardown func(context.Context) error
+}
+
+// Scenario is an interface for executing a sequence of actions in a test scenario. Actions can be added
+// by the relevant struct implementing the interface.
+//
+// The methods of the interface have the following intentions:
+// Apply: should apply the actions that are part of the interface
+// Teardown: should remove or undo the actions applied by Apply
+// Plan: should return a human-readable string describing the actions that will be performed
+type Scenario interface {
+	append(a ...*action)
+	Plan() string
+	Apply(ctx context.Context) error
+	Teardown(ctx context.Context) error
+}
+
+// GithubScenario implements the Scenario interface for testing GitHub functionality. At its base GithubScenario
+// provides two top level methods to create GitHub resources namely:
+// * create GitHub Organization, which returns a codehost_scenario Org
+// * create a GitHub User, which returns a codehost_scenario User
+//
+// Further resources can be created by calling methods on the returned Org or User. For instance, since a repository
+// is tied to an organization, one can call org.CreateRepo, which will add an action for a repo to be created in the
+// organization.
+//
+// Calling any action creating method does not immediately create the resource in GitHub. Instead a action is added
+// the list of actions contained in this scenario. Only once Apply() has been called on the Scenario itself will
+// the resources be created on GitHub.
+//
+// Once Apply() is called, all the corresponding resources should be realized on GitHub. To fetch the corresponding
+// GitHub resources once can call Get() on the resources.
+type GithubScenario struct {
```

Comment:
```suggestion
type GitHubScenario struct {
```

Consistent capitalization 🙏